src/deploy/hosting/uploader.ts

```diff
@@ -0,0 +1,268 @@
+    } else if (!this.populateQueue.finished) {
+      return progressMessage(
+        "adding files to version",
+        this.populateQueue.complete * 1000,
+        this.fileCount
+      );
+    } else if (!this.uploadQueue.finished) {
+      return progressMessage(
+        "uploading new files",
+        this.uploadQueue.complete,
+        this.uploadQueue.stats().total
+      );
+    } else {
+      return "upload complete";
+    }
+  }
+
+  async hashHandler(filePath: string): Promise<void> {
+    const stats = fs.statSync(path.resolve(this.public, filePath));
+    const mtime = stats.mtime.getTime();
+    this.sizeMap[filePath] = stats.size;
+    const cached = this.cache[filePath];
+    if (cached && cached.mtime === mtime) {
+      this.cacheNew.set(filePath, cached);
+      this.addHash(filePath, cached.hash);
+      return;
+    }
+
+    const fstream = this.zipStream(filePath);
+    const hash = crypto.createHash("sha256");
+
+    fstream.pipe(hash);
+
+    return new Promise((resolve, reject) => {
+      fstream.on("end", () => {
+        const hashVal = hash.read().toString("hex");
+        this.cacheNew.set(filePath, { mtime: mtime, hash: hashVal });
+        this.addHash(filePath, hashVal);
+        resolve();
+      });
+
+      fstream.on("error", reject);
+    });
+  }
+
+  addHash(filePath: string, hash: string): void {
+    this.hashMap[hash] = filePath;
+    this.pathMap[filePath] = hash;
+
+    this.populateBatch["/" + filePath] = hash;
+
+    const curBatchSize = size(this.populateBatch);
+    if (curBatchSize > 0 && curBatchSize % this.populateBatchSize === 0) {
+      this.queuePopulate();
+    }
+  }
+
+  queuePopulate(): void {
+    const pop = this.populateBatch;
+    this.populateQueue.add(pop);
+    this.populateBatch = {};
+    this.populateQueue.process();
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async populateHandler(batch: any): Promise<void> {
+    // wait for any existing populate calls to finish before proceeding
+    const res = await this.hashClient.post<
+      unknown,
+      { uploadUrl: string; uploadRequiredHashes: string[] }
+    >(`/${this.version}:populateFiles`, { files: batch });
+    this.uploadUrl = res.body.uploadUrl;
+    this.uploadClient = new Client({ urlPrefix: this.uploadUrl, auth: true });
+    this.addUploads(res.body.uploadRequiredHashes || []);
+  }
+
+  addUploads(hashes: string[]): void {
+    for (const hash of hashes) {
+      this.uploadQueue.add(hash);
+    }
+    this.uploadQueue.process();
+  }
+
+  async uploadHandler(toUpload: string): Promise<void> {
+    if (!this.uploadClient) {
+      throw new FirebaseError("No upload client available.", { exit: 2 });
+    }
+    const controller = new AbortController();
+    const timeout = setTimeout(() => {
+      controller.abort();
+    }, this.uploadTimeout(this.hashMap[toUpload]));
+    const res = await this.uploadClient.request({
```

Comment:
Did you forget to pass in `controller`?

Also, side topic: can we just add `timeout` as an option to `apiv2`? This seems to be a common enough case and while having the flexibility of `controller` is nice, it seems boilerplate-y and prone to errors like forgetting to pass it in, not clearing timeouts, etc. etc.

Reply:
I did, thanks. I didn't know how many use-cases we had for timeouts yet, so I haven't added it to `apiv2`. I'm still back and forth on it a bit, but the case is a little stronger. Not sure how many other spots in the codebase deal with timeouts (I've only seen 2-3, and it's pretty non-trivial logic handling it, usually, which may benefit from more custom solutions like this rather than a single setting)